### PR TITLE
fix bug check nonexist authfile

### LINKED ISSUE #4328: Podman doesn't check existance of given --authfile
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

**Steps to reproduce the issue:**

1. `podman run --auth=/tmp/nonexist --rm alpine sh`

**Describe the results you received:**
Podman starts container as usual

**Describe the results you expected:**
Podman should fail because `/tmp/nonexist ` doesn't exist

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
podman version 1.6.1
```

**Output of `podman info --debug`:**

```
  debug:
  compiler: gc
  git commit: ""
  go version: go1.12.9
  podman version: 1.6.1
host:
  BuildahVersion: 1.11.2
  CgroupVersion: v1
  Conmon:
    package: conmon-2.0.1-1.fc30.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.1, commit: 4346fbe0b2634b05857973bdf663598081240374'
  Distribution:
    distribution: fedora
    version: "30"
  MemFree: 1473187840
  MemTotal: 25091493888
  OCIRuntime:
    package: runc-1.0.0-95.rc9.gitc1485a1.fc30.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc9+dev
      commit: 0840802d505e28d3adcbc7a275cbeadbdf64ddec
      spec: 1.0.1-dev
  SwapFree: 11110969344
  SwapTotal: 12645822464
  arch: amd64
  cpus: 4
  eventlogger: journald
  hostname: sshnaidm.remote
  kernel: 5.3.6-200.fc30.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /usr/bin/slirp4netns
    Package: slirp4netns-0.4.0-4.git19d199a.fc30.x86_64
    Version: |-
      slirp4netns version 0.4.0-beta.2
      commit: 19d199a6ca424fcf9516320a327cedad85cf4dfb
  uptime: 32h 36m 28.81s (Approximately 1.33 days)
registries:
  blocked: null
  insecure: null
  search:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ConfigFile: /home/sshnaidm/.config/containers/storage.conf
  ContainerStore:
    number: 4
  GraphDriverName: overlay
  GraphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-0.6.4-2.fc30.x86_64
      Version: |-
        fusermount3 version: 3.6.2
        fuse-overlayfs: version 0.6.4
        FUSE library version 3.6.2
        using FUSE kernel interface version 7.29
  GraphRoot: /home/sshnaidm/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 21
  RunRoot: /tmp/1000
  VolumePath: /home/sshnaidm/.local/share/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.6.1-2.fc30.x86_64
```

### PR DESCRIPTION
**Pull request by QiWang19**
For podman command(except login), if authfile does not exist returns error.

close #4328 
Signed-off-by: Qi Wang <qiwan@redhat.com>

### CODE REVIEW

**Comment by QiWang19** [comment]
@TomSweeneyRedHat @rhatdan PTAL

File: cmd/podman/shared/funcs.go [hunk 1]
```diff
@@ -8,15 +8,19 @@ import (
...
 func GetAuthFile(authfile string) string {
-	if authfile != "" {
-		return authfile
+	if authfile == "" {
+		authfile = os.Getenv("REGISTRY_AUTH_FILE")
 	}
 
-	authfile = os.Getenv("REGISTRY_AUTH_FILE")
 	if authfile != "" {
+		if _, err := os.Stat(authfile); err != nil {
```
Comment (TomSweeneyRedHat) [inline cmd/podman/shared/funcs.go]: I don't think we need the first if `if authfile != "" {` anymore as we check the inverse up top.

File: test/e2e/create_test.go [hunk 2]
```diff
@@ -241,4 +241,10 @@ var _ = Describe("Podman create", func() {
 		session.WaitWithDefaultTimeout()
 		Expect(session.ExitCode()).To((Equal(0)))
 	})
+
+	FIt("podman create --authfile with nonexist authfile", func() {
+		session := podmanTest.PodmanNoCache([]string{"create", "--authfile", "/tmp/nonexist", "--name=foo", "nginx"})
+		session.WaitWithDefaultTimeout()
+		Expect(session).To(Not(Equal(0)))
+	})
```
Comment (TomSweeneyRedHat) [inline test/e2e/create_test.go]: as painful as it will be, it would be good to add a similar test to each of the other commands.

**Comment by mheon** [comment]
/approve
Change LGTM, but CI is broken

File: cmd/podman/shared/funcs.go (hunk 1 above)
Comment (QiWang19) [reply cmd/podman/shared/funcs.go]: I need if authfile != ""  , os.GetEnv("REGISTRY_FILE_PATH) can still return ""

Comment (TomSweeneyRedHat) [reply cmd/podman/shared/funcs.go]: Oh shoot, yep, you're right.  Never mind!

File: test/e2e/search_test.go [hunk 3]
```diff
@@ -391,4 +391,11 @@ registries = ['{{.Host}}:{{.Port}}']`
 		// cleanup
 		resetRegistriesConfigEnv()
 	})
+
+	// search should fail with nonexist authfile
+	It("podman search fail with nonexist --authfile", func() {
+		search := podmanTest.Podman([]string{"search", "--authfile", "/tmp/nonexist", "alpine"})
+		search.WaitWithDefaultTimeout()
+		Expect(search.ExitCode()).To(Not(Equal(0)))
+	})
```
Comment (TomSweeneyRedHat) [inline test/e2e/search_test.go]: can you add tests for the push, run and play-kube commands?  Are any changes needed for cmd/podman/run.go or cmd/podman/play_kube.go too?  I would suspect so.

**Comment by rhatdan** [comment]
Can't we just do the shared.GetAuthFile() in the options handling, so we check if the environment variable is set and then we don't need to change all this code?

**Comment by QiWang19** [comment]
this will fail if I don't have default  /run/user/1000/containers/auth.json, should we consider this situation that ? I think CI fails for this reason
```
podman search alpine
ERRO[0000] error checking authfile path /run/user/1000/containers/auth.json: stat /run/user/1000/containers/auth.json: no such file or directory 

```

File: cmd/podman/build.go [hunk 4]
```diff
@@ -155,6 +155,10 @@ func buildCmd(c *cliconfig.BuildValues) error {
 		}
 	}
 
+	if _, err := os.Stat(c.BudResults.Authfile); err != nil {
```
Comment (TomSweeneyRedHat) [inline cmd/podman/build.go]: I think the problem with the test is you first need to check if the Authfile has been set.  This gets a little wonky due to the Cobra includes, but I think you can check that with `c.Flag("authfile").Changed`.   You'll need to play/test that making sure that gets tripped appropriately as you pass the auth-file option into the build command.

Comment (QiWang19) [reply cmd/podman/build.go]: this `c.Flag("authfile").Changed`only works for those check that directory runs os.stats under /cmd.
maybe I should get rid of shared.GetSystemContext, just `c.Flag("authfile").Changed`and os.stats for each cmd

**Comment by rhatdan** [comment]
I would like to remove the /run/user... stuff from defaults and allow containers/image to handle it, especially if we move to Keyrings as the default, Then we can rely on C/I to specify the default localtion.

**Comment by TomSweeneyRedHat** [comment]
lint is not your friend @QiWang19

**Comment by rhatdan** [comment]
LGTM

File: cmd/podman/play_kube.go [hunk 5]
```diff
@@ -40,7 +42,7 @@ func init() {
 	flags.BoolVarP(&playKubeCommand.Quiet, "quiet", "q", false, "Suppress output information when pulling images")
 	// Disabled flags for the remote client
 	if !remote {
-		flags.StringVar(&playKubeCommand.Authfile, "authfile", shared.GetAuthFile(""), "Path of the authentication file. Use REGISTRY_AUTH_FILE environment variable to override")
```
Comment (TomSweeneyRedHat) [inline cmd/podman/play_kube.go]: is shared.GetAuthFile being used elsewhere?  If not, did you remove it?  Might answer this question the further I get down in the review here.

File: cmd/podman/runlabel.go [hunk 6]
```diff
@@ -97,6 +98,12 @@ func runlabelCmd(c *cliconfig.RunlabelValues) error {
 	}
 	defer runtime.DeferredShutdown(false)
 
+	if c.Authfile != "" {
+		if _, err := os.Stat(c.Authfile); err != nil {
+			return errors.Wrapf(err, "error getting authfile %s", c.Authfile)
+		}
+	}
+
```
Comment (TomSweeneyRedHat) [inline cmd/podman/runlabel.go]: We seem to be doing this in many of the commands.  Should we make a little utility function in cmd/podman/utils.go?

File: test/e2e/search_test.go (hunk 3 above)
Comment (TomSweeneyRedHat) [reply test/e2e/search_test.go]: @QiWang19 can you answer this one please?

File: cmd/podman/play_kube.go (hunk 5 above)
Comment (QiWang19) [reply cmd/podman/play_kube.go]: shared.GetAuthFile("") has been deleted

File: test/e2e/search_test.go (hunk 3 above)
Comment (QiWang19) [reply test/e2e/search_test.go]: tests were added

File: cmd/podman/runlabel.go (hunk 6 above)
Comment (QiWang19) [reply cmd/podman/runlabel.go]: I think I should add this kind of function in buildah and vendor from there. And add check nonexist authfile in buildah, too.

File: cmd/podman/play_kube.go (hunk 5 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/play_kube.go]: I don't see the removal of that function form cmd/podman/shared/funcs.go in this PR....

File: cmd/podman/runlabel.go (hunk 6 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/runlabel.go]: That's fine by me if that works better.

File: test/e2e/search_test.go (hunk 3 above)
Comment (TomSweeneyRedHat) [reply test/e2e/search_test.go]: I don't see a change to test/e2e/run_test.go, test/e2e/push_test.go or test/e2e/play_kube_test.go in this pr....

File: test/e2e/play_kube_test.go [hunk 7]
```diff
@@ -209,6 +209,16 @@ var _ = Describe("Podman generate kube", func() {
 		processTestResult(f)
 	})
 
+	It("podman play kube fail with nonexist authfile", func() {
```
Comment (QiWang19) [inline test/e2e/play_kube_test.go]: @TomSweeneyRedHat these tests for play_kube_test.go

File: test/e2e/search_test.go (hunk 3 above)
Comment (QiWang19) [reply test/e2e/search_test.go]: push tests are in login_logout_test.go. I will add tests to run_test.go

File: cmd/podman/shared/funcs.go [hunk 8]
```diff
@@ -6,24 +6,19 @@ import (
...
 	"strings"
 
-	"github.com/containers/libpod/pkg/util"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/libpod/libpod/image"
 	"github.com/google/shlex"
+	"github.com/pkg/errors"
 )
 
-func GetAuthFile(authfile string) string {
```
Comment (QiWang19) [inline cmd/podman/shared/funcs.go]: @TomSweeneyRedHat GetAuthFile has been removed here in cmd/podman/shared/funcs.go

Comment (TomSweeneyRedHat) [reply cmd/podman/shared/funcs.go]: Ugh, sorry, must have been tired last night when reviewing.  I swore I searched fro funcs.go and didn't find any instances in the review.  Thanks.

File: test/e2e/play_kube_test.go (hunk 7 above)
Comment (TomSweeneyRedHat) [reply test/e2e/play_kube_test.go]: kk, thx.

File: test/e2e/search_test.go (hunk 3 above)
Comment (TomSweeneyRedHat) [reply test/e2e/search_test.go]: :+1:  thx, I'd rather have an additional test in push_test.go, but I won't hold this up further for that.

**Comment by TomSweeneyRedHat** [comment]
LGTM
Thanks for sticking at it on this PR @QiWang19 !

**Comment by TomSweeneyRedHat** [comment]
and green test buttons if someone wants to LGTM it and merge.

**Comment by QiWang19** [comment]
> and green test buttons if someone wants to LGTM it and merge.

merged this PR after buildah PR https://github.com/containers/buildah/pull/1967 got merged so I can vendor in?

**Comment by TomSweeneyRedHat** [comment]
@QiWang19 I'm a bit confused by your last comment.  Do you want to merge this after containers/buildah#1967 gets merged?  Then you want to vendor Buildah into Podman once that's in?

**Comment by QiWang19** [comment]
@TomSweeneyRedHat yes https://github.com/containers/libpod/pull/4337#discussion_r342794897 if I should add this utility function. That function will be used both in buildah and podman.

**Pull request merged**